### ISSUE TITLE
Cant modify data in hidden field before insert

### ISSUE THREAD

**Issue by odesey** [aldeed/meteor-autoform#949]
I hope this isn't a duplicate question, I have searched and couldn't find anything regarding the issue I am having.

So I have a schema with a few hidden fields, here is the relevant code:

````
Organization.attachSchema(new SimpleSchema({
  name: {
    type: String,
    label: "Organization Name",
    max: 200
  },
  address: {
      type: Object
    },
    'address.street': {
      type: String
    },
    'address.street2': {
      type: String,
      optional: true
    },
    'address.city': {
      type: String
    },
    'address.state': {
      type: String,
      allowedValues: ["AL","AK","AZ","AR","CA","CO","CT","DE","FL","GA","HI","ID","IL","IN","IA","KS","KY","LA","ME","MD","MA","MI","MN","MS","MO","MT","NE","NV","NH","NJ","NM","NY","NC","ND","OH","OK","OR","PA","RI","SC","SD","TN","TX","UT","VT","VA","WA","WV","WI","WY"],
      autoform: {
        afFieldInput: {
          firstOption: "(Select a State)"
        }
      }
    },
    'address.postalCode': {
      type: String,
      regEx: SimpleSchema.RegEx.ZipCode,
      optional: false,
      label: "ZIP"
  },
  location: {
    type: Object,
    autoform: {
      type: "hidden",
      label: false,
      // omit: true
    }
  },
    'location.latitude': {
      type: Number,
      defaultValue: 0.10,
      decimal: true
    },
    'location.longitude': {
      type: Number,
      defaultValue: 0.10,
      decimal: true
    }

}));
````

and the accompanying autoform js:

````
var hooksObject = {
  before: {
    insert: function(doc) {
      console.log(doc) //doc.location object does not exit at this point
      myGeoLocationFunction(doc);
      return doc;
    }
  }
};


AutoForm.hooks({
  orgNew: hooksObject
});
````

The problem is that I am geolocating the address entered before it is stored in the database, however the ````doc.location:```` property for my object does not exist in the ````before insert```` block but it is stored in the DB with the default value.

Am I doing something wrong here?

Thanks.

**Comment by st3phan** [comment]
You are not updating the location property in you doc object. If `myGeoLocationFunction` is returning something try this:

````
var hooksObject = {
  before: {
    insert: function(doc) {
      doc.location = myGeoLocationFunction();
      return doc;
    }
  }
};
````

**Comment by odesey** [comment]
Thanks for the response.

So I tried your suggestion and it still doesn't work. The geo-location is happening on the server side (actually its from the same author of the auto-form package ````aldeed:geocoder````). On the server I do a ````console.log(doc)```` and I see the location attribute and the correct latitude and longitude data. However when I look in the database ````meteor mongo```` the record has the default value even thought I am returning the doc object after inserting the location data.

The real question is, why is the location attribute missing even with the default data assigned at the Schema level?

**Comment by st3phan** [comment]
That's because `aldeed:geocoder` works server-side only and autoform hooks are client-side. So there's the problem. You could change your form type to 'method' and insert the document manually inside your own method. Inside that method you can call the geocoder method.

**Comment by odesey** [comment]
The issue has nothing to do with server vs client. The issue is that I simply cannot change the default values before they are inserted into the database. I have tried the following:

1 - A method call like you suggested for the form type
2 - A ````Meteor.call('myGeoLocateFunction', doc, function(error, result)```` and changing the value      both on the client and on the server
3 - A client side ````geoLocationFunction(doc)````
4 - Simply putting ````doc.location = {}```` followed by ````doc.location.latitude = 123456````

All result with the property value being updated BUT inserted into the database with the default value.

Hopefully that explains it better.

**Comment by st3phan** [comment]
4 should work. You must be doing something wrong.

Here's a meteorpad that does this and works. Check the console.
http://app-d5fqvggi.meteorpad.com/

Meteorpad with editor:
http://meteorpad.com/pad/oQqcWWrNyCgH3FcMa/Autoform%20before%20hook%20test

**Comment by odesey** [comment]
First thanks for your help, greatly appreciated! I managed to figure out what was the issue based on your feedback. Here is why each option I tried was failing:

1 - A method call like you suggested for the form type
1R - The reason the method was not working was because I was modifying the doc on the server and sending it back to the client instead of just doing the insert on the server.....

2 - A ````Meteor.call('myGeoLocateFunction', doc, function(error, result)```` and changing the value  both on the client and on the server
2R - So today I learned what synchronous and asynchronous callbacks are ;-). I made the call to the server to get the geolocation data and the server did get the data and add it to the doc object. Problem was by the time it returned the doc my code on the client continued to execute and saved the client copy of the doc to the database before my callback was complete.

3 - A client side ````geoLocationFunction(doc)````
3R - see 4R

4 - Simply putting ````doc.location = {}```` followed by ````doc.location.latitude = 123456````
4R - I put this code inside my ````Meteor.call```` code block, needless to say I wont make this mistake again anytime soon!

So in the end I used a method like you suggested but this time instead of sending the modified document back to the client, I saved it on the server.
